File: app/src/main/java/com/shrikantbadwaik/tmdb/data/di/module/ApplicationModule.java
package com.shrikantbadwaik.tmdb.data.di.module;

import com.shrikantbadwaik.tmdb.data.repository.Repository;
import com.shrikantbadwaik.tmdb.data.repository.RepositoryImpl;
import com.shrikantbadwaik.tmdb.domain.helper.rx.AppSchedulerProvider;
import com.shrikantbadwaik.tmdb.domain.helper.rx.SchedulerProvider;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module(includes = {NetworkModule.class, ViewModelModule.class})
public class ApplicationModule {

    @Provides
    @Singleton
    public Repository provideRepository(RepositoryImpl repository) {
        return repository;
    }

    @Provides
    @Singleton
    SchedulerProvider provideScheduler() {
        return new AppSchedulerProvider();
    }
}

File: app/src/main/java/com/shrikantbadwaik/tmdb/data/di/module/ApplicationModule.kt
package com.shrikantbadwaik.tmdb.data.di.module

import com.shrikantbadwaik.tmdb.data.repository.Repository
import com.shrikantbadwaik.tmdb.data.repository.RepositoryImpl
import com.shrikantbadwaik.tmdb.domain.helper.rx.SchedulerProvider
import com.shrikantbadwaik.tmdb.domain.helper.rx.SchedulerProviderImpl
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module(includes = [NetworkModule::class, ViewModelBuilder::class])
class ApplicationModule {
    @Provides
    @Singleton
    fun repository(repository: RepositoryImpl): Repository = repository

    @Provides
    @Singleton
    fun schedulerProvider(): SchedulerProvider = SchedulerProviderImpl()
}